packages/opentelemetry-plugin-document-load/src/documentLoad.ts

```diff
@@ -0,0 +1,230 @@
+
+  /**
+   *
+   * @param config
+   */
+  constructor(config: PluginConfig = {}) {
+    super();
+    this._onDocumentLoaded = this._onDocumentLoaded.bind(this);
+    this._config = config;
+  }
+
+  /**
+   * callback to be executed when page is loaded
+   */
+  private _onDocumentLoaded() {
+    this._collectPerformance();
+  }
+
+  /**
+   * Helper function for starting an event
+   * @param span
+   * @param performanceName name of performance entry for time start
+   * @param entries
+   */
+  private _addSpanEvent(
+    span: Span,
+    performanceName: string,
+    entries: PerformanceEntries
+  ): Span | undefined {
+    if (
+      hasKey(entries, performanceName) &&
+      typeof entries[performanceName] === 'number'
+    ) {
+      span.addEvent(performanceName, undefined, entries[performanceName]);
+      return span;
+    }
+    return undefined;
+  }
+
+  /**
+   * Collects information about performance and creates appropriate spans
+   */
+  private _collectPerformance() {
+    const entries = this._getEntries();
+
+    const rootSpan = this._startSpan(
+      AttributeNames.DOCUMENT_LOAD,
+      PTN.FETCH_START,
+      entries
+    );
+    if (!rootSpan) {
+      return;
+    }
+    const fetchSpan = this._startSpan(
+      AttributeNames.DOCUMENT_FETCH,
+      PTN.FETCH_START,
+      entries,
+      {
+        parent: rootSpan,
+      }
+    );
+    if (fetchSpan) {
+      this._addSpanEvent(fetchSpan, PTN.DOMAIN_LOOKUP_START, entries);
+      this._addSpanEvent(fetchSpan, PTN.DOMAIN_LOOKUP_END, entries);
+      this._addSpanEvent(fetchSpan, PTN.CONNECT_START, entries);
+      this._addSpanEvent(fetchSpan, PTN.SECURE_CONNECTION_START, entries);
+      this._addSpanEvent(fetchSpan, PTN.CONNECT_END, entries);
+      this._addSpanEvent(fetchSpan, PTN.REQUEST_START, entries);
+      this._addSpanEvent(fetchSpan, PTN.RESPONSE_START, entries);
+
+      this._endSpan(fetchSpan, PTN.RESPONSE_END, entries);
+    }
+
+    this._addSpanEvent(rootSpan, PTN.UNLOAD_EVENT_START, entries);
+    this._addSpanEvent(rootSpan, PTN.UNLOAD_EVENT_END, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_INTERACTIVE, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_CONTENT_LOADED_EVENT_START, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_CONTENT_LOADED_EVENT_END, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_COMPLETE, entries);
+    this._addSpanEvent(rootSpan, PTN.LOAD_EVENT_START, entries);
+
+    this._endSpan(rootSpan, PTN.LOAD_EVENT_END, entries);
+  }
+
+  /**
+   * Helper function for ending span
+   * @param span
+   * @param performanceName name of performance entry for time end
+   * @param entries
+   */
+  private _endSpan(
+    span: Span | undefined,
+    performanceName: string,
+    entries: PerformanceEntries
+  ) {
+    // span can be undefined when entries are missing the certain performance - the span will not be created
+    if (typeof span !== 'undefined' && hasKey(entries, performanceName)) {
```

Comment:
nit: `'undefined'` should be made a constant (probably outside this plugin), (e.g. `Constants.Undefined`), so that this can be minified better

Reply:
Would it work to just do `span !== undefined` instead?

I would hope that smart minifier should be able to hoist out shared constants

Reply:
(Though I don't know if they actually do!) I guess to me it feels a little weird to try to do that optimization if it could be done better automatically

Reply:
You are right, I think `!== undefined` would work here. I think `typeof` is only needed for global stuff.

For the string constants, uglify won't do any hoisting here, but maybe other bundlers do?

Reply:
I suspect the Closure compiler can though I have not specifically verified it (see 
https://developers.google.com/closure/compiler/docs/api-tutorial3#better)

There is actually a really cool project called [Tsickle](https://github.com/angular/tsickle) that can transform TypeScript code into Closure compiler type annotations and then have it feed into the advanced compilation mode for Closure. I would like to eventually see us be able to distribute `<script>` tag files for the browser that have been fully minified for basic things like the document load instrumentation. That could help people who aren't interested in custom spans but just want a minimally-invasive way of instrumenting their page loads. Same could go for people who just want page load + out of the box user interaction tracing.

Reply:
with regards to
```
Alternatively, to get the intended this scoping, does TS complain about making the class member function an arrow function? Or would this go against style guidelines?
```
this if fine if you add listener in the same function but here the `_onDocumentLoaded` is being added in `patch` and removed in `unpatch` function - that's why the context needs to be bind before so you remove listener correctly

Reply:
With regards to `micro optimisation` - I think this should be a part of uglifier not the code itself.
And `typeof someVariable === 'undefined'` vs `someVariable === undefined` .

First example will never raise an error the second will raise a `ReferenceError` if such thing isn't declared. There used to be also some extra error with this in IE too. To keep consistency I'm choosing the first option always to avoid possible errors as `typeof` will never have such issues.